### PR TITLE
470.57.02

### PR DESCRIPTION
**Pull request by leviport**
https://www.nvidia.com/Download/driverResults.aspx/177145/en-us

### CODE REVIEW

**Comment by cybik** [comment]
# Test Writeup

## Issue

Using the latest 470.57.02 drivers, 2 out of 3 of my monitors exhibit a strange backlight behaviour where the backlight level seems to wave, essentially causing sort of a light flickering that becomes bothersome within a minute.

## Hardware configuration

* Non-Pop hardware
* i7 8700k
* nVidia RTX3060Ti
* Pop 21.04 beta
* nVidia 470.57.02
* 3 screens, two of which Adaptive sync enabled (One true GSync, one Forced GSync)
  * The third screen not being GSync and not exhibiting this behaviour, it's possible the issue is related to adaptive sync.
  * Disabling Freesync on the ultrawide fixes the issue - meaning Adaptive Sync is broken at the driver level

## Software configuration

### Behaviour tested on
* SDDM + KDE/KF5 + KWin, without and with modesetting (boot option)
* GDM3 + Gnome Shell, without modesetting (boot option, or specifically lack thereof)

### Kernels tested
* xanmod1-calcule 5.12.13

## Miscelaneous details

* All the monitors showing the issue are connected over DisplayPort; the non-issue monitor is over HDMI
* The monitors with issue are set to a >1080p resolution

**Comment by jacobgkau** [comment]
@cybik Can you please open the NVIDIA X Server Settings app, navigate to X Server Display Configuration -> Advanced, and enable the Force Full Composition Pipeline setting, then click Apply? I believe @bflanagin found that would stop the display flickering on G-Sync/FreeSync displays.

**Comment by cybik** [comment]
Damn. It does.

Still doesn't explain why it broke in the upgrade but I'll take it.

File: debian/libnvidia-common-470.install [hunk 1]
```diff
@@ -0,0 +1,2 @@
+NVIDIA-Linux/nvidia-application-profiles-470.57.02-key-documentation        usr/share/nvidia
+NVIDIA-Linux/nvidia-application-profiles-470.57.02-rc                       usr/share/nvidia
```
Comment (leviport) [inline debian/libnvidia-common-470.install:2]: It looks like these two files are not being created, and that's what's causing the adaptive sync flickering issue. @bflanagin found that adding `/usr/share/nvidia/nvidia-application-profiles-470.57.02-key-documentation` and `/usr/share/nvidia/nvidia-application-profiles-470.57.02-rc` manually caused adaptive sync to work without flickering. I found them in the .run file when I extracted it, so my guess is that I messed something up in the debian/control file. I'm just having a hard time figuring out what exactly that might be. @mmstick do you see anything obvious?

Comment (cybik) [reply debian/libnvidia-common-470.install:2]: Right now this is what I see:

```
cybik@localhost > /usr/share/nvidia > ls -lh
total 20K
-rw-r--r-- 1 root root 7,0K mai 13 15:12 nvidia-application-profiles-465.31-key-documentation
-rw-r--r-- 1 root root 8,7K mai 13 15:12 nvidia-application-profiles-465.31-rc
```

And you're right to say that 470 doesn't have them. I have common-465 and common-470 installed at the same time right now and only 465 has them.

Comment (leviport) [reply debian/libnvidia-common-470.install:2]: Yep, and when I uninstalled libnvidia-common-465, then purged and reinstalled everything NVIDIA, I was left with no `/usr/share/nvidia/` directory at all. It's left me very confused.

Comment (ids1024) [reply debian/libnvidia-common-470.install:2]: Weird. The package seems to include those files when I build it locally with `fakeroot make -f debian/rules clean binary`, but the version on `http://apt.pop-os.org` built from the same commit doesn't seem to have the files.

Comment (leviport) [reply debian/libnvidia-common-470.install:2]: Maybe I just need to jiggle the handle. I'll try pushing an empty commit to make it rebuild.

**Comment by cybik** [comment]
Profiles still missing in the latest `libnvidia-common-470` package. Something's odd af.

**Comment by leviport** [comment]
Yeah I'm seeing that too. Interestingly, if I install the 470.57.02 driver in Ubuntu 21.04, the profiles are added correctly. I'm currently investigating what the differences might be between this package and Ubuntu's.

**Comment by mmstick** [comment]
I have used Meld to sync our debian packaging with Ubuntu's, preserving our modeset and xorg configuration changes. It added a few libs that weren't getting included, though I didn't see any changes specific to those profiles.

**Comment by cybik** [comment]
@mmstick *still* the common package is missing the magic files.

**Comment by bflanagin** [comment]
Looking at previous libnvidia-common deb files the last version that contained the files in question is libnvidia-common-465_465.31-1pop0~1623777959~21.04~a3ca7f3_all.deb

**Comment by mmstick** [comment]
The files are definitely there when I build locally with `sbuild`.

**Comment by cybik** [comment]
I am obviously talking out of my proverbial backside, but either

* the .run you're using is a cached, broken version that does *not* contain the files
* the files are removed after decompression, and thus cannot be grabbed
* some other shenanigans are afoot

I've little experience with debian packaging, but these would be my first impressions.

**Comment by mmstick** [comment]
sbuild is what the build servers use for packaging, and they create pristine chroots with no access to the network for building packages inside, and any changes made in these chroots are of course file system overlays that are discarded after one use.

**Comment by mmstick** [comment]
I'm wondering if Launchpad would have the same issue. We could merge to master and then check if the staging packages in the proposed PPA have the missing files. In general, debian packaging would error if those files listed in the `.install` file were missing when it was packaging.

**Comment by ids1024** [comment]
Looking at the files manually:

http://apt.pop-os.org/staging/470.57.02/pool/hirsute/nvidia-graphics-drivers/nvidia-graphics-drivers-470_470.57.02-1pop0~1626951027~21.04~6df8618.tar.xz
http://apt.pop-os.org/staging/470.57.02/pool/hirsute/nvidia-graphics-drivers/libnvidia-common-470_470.57.02-1pop0~1626951027~21.04~6df8618_all.deb

The source archive, and manual builds from it, seems fine. The `.deb` is missing the files according to `dpkg -c`.

**Comment by leviport** [comment]
So if I'm understanding correctly, the `libnvidia-common-470` package is what is supposed to copy those files upon install, right? I'm not sure if this is significant, but if I `apt download libnvidia-common-470` on both Pop 21.04 and Ubuntu 21.04, then `dpkg -I` the .deb file that downloads, this is the difference in output I get between Pop and Ubuntu. The size difference between the two is the main thing I thought was suspicious.


I'm ok with building it in proposed on Launchpad if we're out of leads, but I think libnvidia-common-470 might be worth digging into.

**Comment by leviport** [comment]
@mmstick feel free to merge if you want to try building on Launchpad. I'd add an approval, but it's my PR. Just no releasing until we know those files are being added correctly, of course.

**Comment by mmstick** [comment]
Ping me if it's confirmed we have it working. Else I'll roll back and re-open a PR.

**Comment by leviport** [comment]
So far I'm seeing the magic files being added where they're supposed to be on my machine. Now we're investigating some possible weirdness with switchable graphics machines. It's probably nothing, but we want to be make sure.

**Comment by bflanagin** [comment]
When switching to Hybrid from Nvidia I'm being thrown back into Integrated mode. 

From journalctl:

```
Jul 22 11:33:29 flanagins-gazelle systemd-udevd[505]: nvidia: Process '/sbin/modprobe nvidia-modeset' failed with exit code 1.
Jul 22 11:33:29 flanagins-gazelle kernel: nvidia-nvlink: Unregistered the Nvlink Core, major device number 511
Jul 22 11:33:29 flanagins-gazelle kernel: nvidia: probe of 0000:01:00.0 failed with error -1
Jul 22 11:33:29 flanagins-gazelle kernel: nvidia 0000:01:00.0: vgaarb: changed VGA decodes: olddecodes=none,decodes=none:owns=none
Jul 22 11:33:29 flanagins-gazelle kernel: nvidia 0000:01:00.0: can't change power state from D3cold to D0 (config space inaccessible)
Jul 22 11:33:29 flanagins-gazelle kernel: nvidia-nvlink: Nvlink Core is being initialized, major device number 511
Jul 22 11:33:29 flanagins-gazelle systemd[1]: nvidia-persistenced.service: Failed with result 'exit-code'.
```

**Comment by mmstick** [comment]
I'm getting `nvidia 0000:01:00.0: can't change power state from D3cold to D0 (config space inaccessible)`

**Comment by mmstick** [comment]
I believe we can fix this by disabling some of the newer udev rules in the Ubuntu packaging

**Comment by mmstick** [comment]
I've pushed a fix for this to master

**Comment by bflanagin** [comment]
D3cold issue is resolved on my end after the push. Everything is looking good, just want to do one or two more tests to be on the safe side.

**Comment by bflanagin** [comment]
Previous tests as well as our G-sync (Adaptive Sync) tests are all passing. The missing files are in place. I believe its ready for release.

**Comment by cybik** [comment]
User confirmation: the latest update packages for 470.57.02 contain the application profile files.

**Pull request merged**